cassiopeia/simulator/BirthDeathFitnessSimulator.py

```diff
@@ -0,0 +1,409 @@
+                )
+                current_lineages.put(
+                    (
+                        death_waiting_time + lineage["total_time"],
+                        unique_id,
+                        {
+                            "id": unique_id,
+                            "birth_scale": lineage["birth_scale"],
+                            "total_time": death_waiting_time
+                            + lineage["total_time"],
+                            "active": False,
+                        },
+                    )
+                )
+                return unique_id + 1
+
+        # Instantiate the implicit root
+        tree = nx.DiGraph()
+        tree.add_node(0)
+        tree.nodes[0]["birth_scale"] = initial_birth_scale
+        tree.nodes[0]["total_time"] = 0
+        current_lineages = PriorityQueue()
+        # Records the nodes that are observed at the end of the experiment
+        observed_nodes = []
+        starting_lineage = {
+            "id": 0,
+            "birth_scale": initial_birth_scale,
+            "total_time": 0,
+            "active": True,
+        }
+
+        # Sample the waiting time until the first division
+        unique_id = 1
+        unique_id = sample_lineage_event(unique_id, starting_lineage)
+
+        # Perform the process until there are no active extant lineages left
+        while not current_lineages.empty():
+            # If number of extant lineages is the stopping criterion, at the
+            # first instance of having n extant tips, stop the experiment
+            # and set the total lineage time for each lineage to be equal to
+            # the minimum, to produce ultrametric trees. Also, the birth_scale
+            # parameter of each leaf is rolled back to equal its parent's.
+            if num_extant:
+                if current_lineages.qsize() == num_extant:
+                    remaining_lineages = []
+                    while not current_lineages.empty():
+                        _, _, lineage = current_lineages.get()
+                        remaining_lineages.append(lineage)
+                    min_total_time = remaining_lineages[0]["total_time"]
+                    for lineage in remaining_lineages:
+                        parent = list(tree.predecessors(lineage["id"]))[0]
+                        tree.edges[parent, lineage["id"]]["weight"] += (
+                            min_total_time - lineage["total_time"]
+                        )
+                        tree.nodes[lineage["id"]]["birth_scale"] = tree.nodes[
+                            parent
+                        ]["birth_scale"]
+                        observed_nodes.append(lineage["id"])
+                    break
+            # Pop the minimum age lineage to simulate forward time
+            _, _, lineage = current_lineages.get()
+            for _ in range(2):
+                unique_id = sample_lineage_event(unique_id, lineage)
```

Comment:
@richardyz98 I would rather we only call `sample_lineage_event(unique_id, lineage)` if the lineage is active. If it is inactive, the function `sample_lineage_event(unique_id, lineage)` does nothing anyway, so the end result is the same. Thus, we might as well put the check here. In `sample_lineage_event(unique_id, lineage)` we can further assert that the lineage is active. This will make the code a lot easier to understand because the invariants will be clearer.